### PR TITLE
Fetch all matching rules for a fact

### PR DESCRIPTION
**Pull request by jinagamvasubabu**
## Background:
Currently there is no feature available to fetch all the matching rules against a fact order by salience.
* This will help me to  find out all the rules available against a fact before creating one.

## How to call ?
```go
engine := engine.NewGruleEngine()
ruleEntries, err := engine.FetchMatchingRules(context.Background(), dctx, kb)
if err != nil {
    panic(err)
}
```
Signature looks more like `Execute` function in `GruleEngine.go`

## Unit Tests?
Yes i have covered the unit tests, let me know if you want me to add more

### CODE REVIEW

File: docs/Tutorial_en.md [hunk 1]
```diff
@@ -235,3 +235,105 @@ fmt.Println(myFact.WhatToSay)
 // this should prints
 // Lets Say "Hello Grule"
 ```
+
```
Comment (newm4n) [inline docs/Tutorial_en.md:238]: I think, this documentation portion is too advanced for those who just start learning how to use Grule. 
I would advise you to move this portion to a separate .md file and have it linked from `FAQ_en.md` under a new FAQ entry. eg. Q : How can I test which of rules I define is valid for a given Facts... A : You can use engine.FetchMatchingRule function as [documented here](link). Something like that... :)

File: examples/MatchingRules_test.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
+package examples
```
Comment (newm4n) [inline examples/MatchingRules_test.go:1]: What is the difference between this test and the one in `GruleEngine_test.go` ? If its doing the same test, we wont need this one. Usually I put tests in `examples` folder for testing reported bug/issue, and to show the reporter that such bug/issue are no longer appear or not reproducable.

File: engine/GruleEngine.go [hunk 3]
```diff
@@ -206,3 +206,47 @@ func (g *GruleEngine) ExecuteWithContext(ctx context.Context, dataCtx ast.IDataC
 
 	return nil
 }
+
+// FetchMatchingRules function is responsible to fetch all the rules that matches to a fact against all rule entries
+// Returns []*ast.RuleEntry order by salience
+func (g *GruleEngine) FetchMatchingRules(ctx context.Context, dataCtx ast.IDataContext, knowledge *ast.KnowledgeBase) ([]*ast.RuleEntry, error) {
```
[the remark this answers is not in the document]
Comment (jinagamvasubabu) [reply engine/GruleEngine.go:212]: Sure, totally makes sense

**Comment by jinagamvasubabu** [comment]
I have done the changes requested, please let me know if there are any changes required ?

**Pull request merged**